### PR TITLE
Fixes meson for systemd-boot AArch64 cross-build

### PR DESCRIPTION
**Pull request by samueldr**
Note about `cpu_family`:

As documented, it should be `aarch64` for AArch64.

 * https://mesonbuild.com/Reference-tables.html#cpu-families

```
 $ nix eval '((import <nixpkgs> {}).pkgsCross.aarch64-multiplatform.stdenv.targetPlatform.parsed.cpu.family)'
"arm"
```

This explains the new conditional. `family` seems fine for at least
~`x86_64`~, `i686` and `armv7`.

* * *

This should be merged together with #61321; dropping the systemd AArch64 change on one of the PRs.

###### Things done

**This was tested when rebased on nixos-unstable. In the current state of staging, something (glib) doesn't build for at-leat cross-compilation, it looks like.**

- ☑️ Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - ☑️ NixOS
   - 🔲 other Linux distributions
- 🔲 Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- 🔲 Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- 🔲 Tested execution of all binary files (usually in `./result/bin/`)
- 🔲 Determined the impact on package closure size (by running `nix path-info -S` before and after)
- ☑️ Assured whether relevant documentation is up to date
- ☑️ Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc @andir

### CODE REVIEW

**Comment by matthewbauer** [comment]
It looks like it's incorrect for x86_64 as well? ```parsed.cpu.family``` gives x86 but it looks like meson expected x86_64.

**Comment by samueldr** [comment]
Oops! You're right, that's bad sleuthing on my part.

Do you have an alternative to special casing both of those? It seems a bit clunky.

File: pkgs/development/tools/build-managers/meson/default.nix
Comment (flokli) [inline pkgs/development/tools/build-managers/meson/default.nix]: Peeking at `lib/systems/parse.nix`, `targetPackages.stdenv.targetPlatform.parsed.cpu.name` looks more correct in most cases - for example, for `riscv32` and `riscv64` `targetPackages.stdenv.targetPlatform.parsed.cpu.family` is just `riscv`, but `targetPackages.stdenv.targetPlatform.parsed.cpu.name` is correct.

Comment (flokli) [reply pkgs/development/tools/build-managers/meson/default.nix]: This is correct for `x86_64` too - `x86` would need to be special-cased, given `cpu.name` is `i686` there, and while it might work, it's not guaranteed to stay, according to meson docs.

Comment (samueldr) [reply pkgs/development/tools/build-managers/meson/default.nix]: > it's not guaranteed to stay, according to meson docs.

I'm not sure I follow, I don't see anything special about `x86` in their docs, but I don't really know where too look.

Comment (matthewbauer) [reply pkgs/development/tools/build-managers/meson/default.nix]: Yeah I would make the condition something like:

```
if targetPackages.stdenv.targetPlatform.isAarch32 then "arm"
else targetPackages.stdenv.targetPlatform.parsed.cpu.name
```

there's a few more cases like powerpc, riscv, that we probably should handle, but this will take care of most cases.

Comment (flokli) [reply pkgs/development/tools/build-managers/meson/default.nix]: I wanted to suggest using `….cpu.name` for mesons `cpu_family` instead of what we get from `….cpu.family`, as it seemed to fit better.

I did take a closer look:

`….cpu.name` is correct for the following meson CPU families:
 - `aarch64`
 - `riscv32`
 - `riscv64`

`….cpu.name` is wrong for the following meson CPU families
 - 32 bit arm flavours (the armv7*) ones at least
 - `i686` instead of `x86`
 - `powerpc` / `powerpcle` instead of `ppc`
 - `armv7l

I ignored the ones not available in nixpkgs via `pkgsCross`.

In general, this looks very messy, and hard to test / ensure to keep working.

I'd suggest to instead provide a fixed lookup table / lookup attrset in the meson drv, translating from `parsed.cpu.name` to mesons `cpu_family`.

Comment (samueldr) [reply pkgs/development/tools/build-managers/meson/default.nix]: Well, you replied just as I was pushing a new revision, though yeah I think the attrset approach might end up being better considering the different discrepancies.

**Review by Ericson2314 — APPROVED** [review]
This looks good now. Thanks!

**Comment by andir** [comment]
I pulled the latest changes into the systemd PR.

**Comment by samueldr** [comment]
Leaving the PR up as a draft until #61321 is finalized and merged (I *think* if the PR body says it closes #61847 this PR will be closed automatically).

**Comment by flokli** [comment]
@samueldr isn't this relatively independent from each other? Or do we just want to avoid one world rebuild and merge this in together with the systemd bump for that reason?

**Comment by samueldr** [comment]
It's to avoid the rebuild. Though in practice the meson changes could go in. I think it could even go in to master since it should be a no-op for native builds. So, I guess I leave that to @andir depending on how he wants to proceed.

**Comment by andir** [comment]
@samueldr Go ahead an merge it into master. I am all for having changes outside of the systemd PR, if they make sense on their own. I'll happily rebase my PR a few more times if that means we can get changes into nixpkgs incrementally without a big bang change :-)

**Comment by samueldr** [comment]
(Maybe) closing in favor of a PR targeting master.

(Also testing something with regards to changing the branch and force push...)

**Pull request closed without merging**
